# **Lab 2: Indexing and Subsetting (20 points)**

#### **WRITE YOUR NAME HERE**


 **Alberto Pulido**

---

**IMPORTANT**:
- Before starting the lab, save a copy to your google drive by navigating to `File > Save a copy in Drive` <br>
- Then rename the copied file `Lab#_LASTNAME_FIRSTNAME.ipynb` where # is the lab assignment number, LASTNAME is your last name, and FIRSTNAME is your first name.

<br>


**Instructions**: Complete the lab below.

- <font color=red>Tasks</font> indicated in red must be completed to receive credit
- Include all code for your work
- Include markdown cells with written answers explaining your work when prompted
- **NOTE: Your completed lab must be submitted to Canvas as a .ipynb file by the end of the lab session**
---

<br>

<br>






####<font color=red>Organized</font> (1 points)
* Answers should be organized in the following format for organization and readability

<br>

```
# 1.
This will be the code block to answer the first part of a task
```

1. This will be the text answer/explanation (if prompted) for the first task

<br>

```
# 2.
This will be the code block to answer the second part of a task in a separate code cell
```

2. This will be the text answer/explanation (if prompted) for the second task

<br>

Continue this format for the remaining subtasks


<br>

---

<br>






####<font color=red>Task 1</font>  (1 points)

The code below loads and preprocesses the US Census data described in lecture
* Run the code cell
* View the first few rows using `head()` to ensure the data has been properly reformatted for analysis


In [73]:
# create url to census data
census_url <- url("https://www2.census.gov/programs-surveys/popest/tables/2010-2011/state/totals/nst-est2011-01.csv")

# load the census data into memory
census <- read.csv(census_url)

# remove unnecessary columns and rows
nb_row <- nrow(census)
census <- census[-c(1:8, (nb_row - 6):nb_row), -(6:10)]

# overwrite the column names using meaningful column names
names(census) <- c("state", "april2010census", "april2010base", "july2010pop", "july2011pop")

# remove periods from state names
census$state <- gsub('\\.', '', census$state)

# remove commas from populations and convert to numeric
census$april2010census <- as.numeric(gsub(',', '', census$april2010census))
census$april2010base <- as.numeric(gsub(',', '', census$april2010base))
census$july2010pop <- as.numeric(gsub(',', '', census$july2010pop))
census$july2011pop <- as.numeric(gsub(',', '', census$july2011pop))

# remove existing row names
rownames(census) <- NULL

In [3]:
# View the first few rows
head(census)

state      april2010census april2010base july2010pop july2011pop
1 Alabama     4779736         4779735       4785401     4802740   
2 Alaska       710231          710231        714146      722718   
3 Arizona     6392017         6392013       6413158     6482505   
4 Arkansas    2915918         2915921       2921588     2937979   
5 California 37253956        37253956      37338198    37691912   
6 Colorado    5029196         5029196       5047692     5116796

<br>

---

<br>






####<font color=red>Task 2</font>  (4 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. What is the average population across states in July 2010?
2. What is the average population across states in July 2011?
3. Was there an increase or decrease in the average population?

Hint: Use `mean()`

In [10]:
mean(census$july2010pop)
mean(census$july2011pop)
mean(census$july2011pop) - mean(census$july2010pop)


[1] 6065298

[1] 6109645

[1] 44347.02

**1. The average population across states in July 2010 was 6,065,298 people.**
**2. The average population across states in July 2011 was 6,109,645 people.**
**3. There was an increase in population from July 2010 to July 2011.**

<br>

---

<br>





####<font color=red>Task 3</font>  (2 points)
Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.


1. What is the largest population in July 2010? [Hint: Use `max()`]

The function `which.max()` provides you with the index of the maximum. For example, the output of `which.max(c(99, 256, 48))` is 2 since the second entry of the vector, 256, is the maximum.
2. Use `which.max()` to find the state with the largest population.

In [15]:
max(census$july2010pop)
which.max(census$july2010pop)
census$state[which.max(census$july2010pop)]

[1] 37338198

[1] 5

[1] "California"

**1. The largest population is 37,338,198.**
**2. The state with the largest population is California.**

####<font color=red>Task 4</font>  (2 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.
1. Use `which.min()` to find the state with the smallest population in July 2010.
2. What is the population in this state?

In [19]:
which.min(census$july2010pop)
census$state[51]

min(census$july2010pop)

[1] 51

[1] "Wyoming"

[1] 564554

**1. The state with the smallest population in July 2010 is Wyoming.**
**2. The population of the smallest state, Wyoming, is 564,554**

<br>

---

<br>






####<font color=red>Task 5</font>  (4 points)

The function `order()` returns the indexes of a sorted vector. For example, the output of `order(c(99, 256, 48))` is the vector `(3, 1, 2)` since the third number is the smallest (48), the first number is next (99), and the second number is the largest (256).

The example below shows how to sort a single vector:
```
x <- c(99, 256, 48)
x[order(x)]
```
<br>



1. Sort the dataframe `census` by population in July 2010 from smallest to largest and store as another dataframe called `census_sorted` <br>
Hint: `census[order(census$july2010pop), ]`
2. Show the first few lines of `census_sorted` to ensure it is correct
3. Inferring from the sorted dataframe, which areas of the US have the smallest populations? Why?
4. Inferring from the sorted dataframe, which areas of the US have the largest populations? Why?

In [70]:
census_sorted = census[order(census$july2010pop),]
head(census_sorted)
tail(census_sorted)

state                april2010census april2010base july2010pop july2011pop
51 Wyoming              563626          563626        564554      568158     
9  District of Columbia 601723          601723        604912      617996     
46 Vermont              625741          625741        625909      626431     
35 North Dakota         672591          672591        674629      683932     
2  Alaska               710231          710231        714146      722718     
42 South Dakota         814180          814180        816598      824082     
   pop_diff pop_diff.1 pop_diff.2 pop_diff.3
51  3604     3604       3604       3604     
9  13084    13084      13084      13084     
46   522      522        522        522     
35  9303     9303       9303       9303     
2   8572     8572       8572       8572     
42  7484     7484       7484       7484

state        april2010census april2010base july2010pop july2011pop pop_diff
39 Pennsylvania 12702379        12702379      12717722    12742886     25164  
14 Illinois     12830632        12830632      12841980    12869257     27277  
10 Florida      18801310        18801311      18838613    19057542    218929  
33 New York     19378102        19378104      19395206    19465197     69991  
44 Texas        25145561        25145561      25253466    25674681    421215  
5  California   37253956        37253956      37338198    37691912    353714  
   pop_diff.1 pop_diff.2 pop_diff.3
39  25164      25164      25164    
14  27277      27277      27277    
10 218929     218929     218929    
33  69991      69991      69991    
44 421215     421215     421215    
5  353714     353714     353714

**3. The smallest states by population are Wyoming, D.C., Vermont, North Dakota, Alaska, and South Dakota due to using the order fxn above.**
**4. The biggest states by population are Pennsylvania, Illinois, Florida, New York, Texas and California for the same reason listed above.**

####<font color=red>Task 6</font>  (2 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Create a dataframe of states with populations between 1 and 2 million people in July 2010 and store as `census_1to2`.
2. Among states with populations between 1 and 2 million people, which state has the largest population? Use R code to answer this question. [Hint: `which.max()`]

Hint: `census$july2010pop > 1000000 & census$july2010pop < 2000000`

In [71]:
census_1to2 = census$july2010pop > 1000000 & census$july2010pop < 2000000
census_1to2 = census[census_1to2,]
census_1to2$state[which.max(census_1to2$july2010pop)]

[1] "West Virginia"

**2. West Virginia has the biggest population size.**

####<font color=red>Task 7</font>  (4 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Calculate the difference in population for each state between 2010 and 2011. [Hint: population 2011 - population 2010]
2. Store as another variable called `pop_diff` in the `census` dataframe
3. How many states decreased in population between 2010 and 2011? (i.e. How many states have `pop_diff` less than 0?)
4. How many states had population increases less than 10,000 between 2010 and 2011? Hint: `pop_diff` greater than 0 and less than 10,000


**3. 2 states saw a decrease for the conditions listed between 2010 & 2011.**
**4. 13 states saw an increase for the conditions listed between 2010 and 2011. **

In [84]:
pop_diff = (census$july2011pop) - (census$july2010pop)
pop_diff
census = data.frame(census,pop_diff)
census
census_dec = census[pop_diff < 0,]
census_dec
census_inc = census[pop_diff > 0 & pop_diff < 10000,]
census_inc
nrow(census_dec)
nrow(census_inc)

[1]  17339   8572  69347  16391 353714  69104   5211   7343  13084 218929
[11] 103053  11451  13883  27277  26300  12107  12095  22133  29493    809
[21]  42608  32070   -956  34203   8440  14973   7241  12500  19039   1387
[31]  21562  16311  69991  96167   9303   6983  31324  33527  25164  -1226
[41]  42124   7484  45917 421215  41743    522  72651  87088    996  20108
[51]   3604

state                april2010census april2010base july2010pop july2011pop
1  Alabama               4779736         4779735       4785401     4802740   
2  Alaska                 710231          710231        714146      722718   
3  Arizona               6392017         6392013       6413158     6482505   
4  Arkansas              2915918         2915921       2921588     2937979   
5  California           37253956        37253956      37338198    37691912   
6  Colorado              5029196         5029196       5047692     5116796   
7  Connecticut           3574097         3574097       3575498     3580709   
8  Delaware               897934          897934        899792      907135   
9  District of Columbia   601723          601723        604912      617996   
10 Florida              18801310        18801311      18838613    19057542   
11 Georgia               9687653         9687660       9712157     9815210   
12 Hawaii                1360301         1360301       1363359     1374810   
13 Idaho                 1567582         1567582       1571102     1584985   
14 Illinois             12830632        12830632      12841980    12869257   
15 Indiana               6483802         6483800       6490622     6516922   
16 Iowa                  3046355         3046350       3050202     3062309   
17 Kansas                2853118         2853118       2859143     2871238   
18 Kentucky              4339367         4339362       4347223     4369356   
19 Louisiana             4533372         4533372       4545343     4574836   
20 Maine                 1328361         1328361       1327379     1328188   
21 Maryland              5773552         5773552       5785681     5828289   
22 Massachusetts         6547629         6547629       6555466     6587536   
23 Michigan              9883640         9883635       9877143     9876187   
24 Minnesota             5303925         5303925       5310658     5344861   
25 Mississippi           2967297         2967297       2970072     2978512   
26 Missouri              5988927         5988927       5995715     6010688   
27 Montana                989415          989415        990958      998199   
28 Nebraska              1826341         1826341       1830141     1842641   
29 Nevada                2700551         2700551       2704283     2723322   
30 New Hampshire         1316470         1316472       1316807     1318194   
31 New Jersey            8791894         8791894       8799593     8821155   
32 New Mexico            2059179         2059180       2065913     2082224   
33 New York             19378102        19378104      19395206    19465197   
34 North Carolina        9535483         9535475       9560234     9656401   
35 North Dakota           672591          672591        674629      683932   
36 Ohio                 11536504        11536502      11537968    11544951   
37 Oklahoma              3751351         3751354       3760184     3791508   
38 Oregon                3831074         3831074       3838332     3871859   
39 Pennsylvania         12702379        12702379      12717722    12742886   
40 Rhode Island          1052567         1052567       1052528     1051302   
41 South Carolina        4625364         4625364       4637106     4679230   
42 South Dakota           814180          814180        816598      824082   
43 Tennessee             6346105         6346110       6357436     6403353   
44 Texas                25145561        25145561      25253466    25674681   
45 Utah                  2763885         2763885       2775479     2817222   
46 Vermont                625741          625741        625909      626431   
47 Virginia              8001024         8001030       8023953     8096604   
48 Washington            6724540         6724540       6742950     6830038   
49 West Virginia         1852994         1852996       1854368     1855364   
50 Wisconsin             5686986         5686986       5691659     5711767   
51 Wyoming               

state        april2010census april2010base july2010pop july2011pop pop_diff
23 Michigan     9883640         9883635       9877143     9876187      -956   
40 Rhode Island 1052567         1052567       1052528     1051302     -1226   
   pop_diff.1 pop_diff.2 pop_diff.3 pop_diff.4 pop_diff.5 pop_diff.6 pop_diff.7
23  -956       -956       -956       -956       -956       -956       -956     
40 -1226      -1226      -1226      -1226      -1226      -1226      -1226     
   pop_diff.8 pop_diff.9 pop_diff.10
23  -956       -956       -956      
40 -1226      -1226      -1226

state         april2010census april2010base july2010pop july2011pop pop_diff
2  Alaska          710231          710231        714146      722718    8572    
7  Connecticut    3574097         3574097       3575498     3580709    5211    
8  Delaware        897934          897934        899792      907135    7343    
20 Maine          1328361         1328361       1327379     1328188     809    
25 Mississippi    2967297         2967297       2970072     2978512    8440    
27 Montana         989415          989415        990958      998199    7241    
30 New Hampshire  1316470         1316472       1316807     1318194    1387    
35 North Dakota    672591          672591        674629      683932    9303    
36 Ohio          11536504        11536502      11537968    11544951    6983    
42 South Dakota    814180          814180        816598      824082    7484    
46 Vermont         625741          625741        625909      626431     522    
49 West Virginia  1852994         1852996       1854368     1855364     996    
51 Wyoming         563626          563626        564554      568158    3604    
   pop_diff.1 pop_diff.2 pop_diff.3 pop_diff.4 pop_diff.5 pop_diff.6 pop_diff.7
2  8572       8572       8572       8572       8572       8572       8572      
7  5211       5211       5211       5211       5211       5211       5211      
8  7343       7343       7343       7343       7343       7343       7343      
20  809        809        809        809        809        809        809      
25 8440       8440       8440       8440       8440       8440       8440      
27 7241       7241       7241       7241       7241       7241       7241      
30 1387       1387       1387       1387       1387       1387       1387      
35 9303       9303       9303       9303       9303       9303       9303      
36 6983       6983       6983       6983       6983       6983       6983      
42 7484       7484       7484       7484       7484       7484       7484      
46  522        522        522        522        522        522        522      
49  996        996        996        996        996        996        996      
51 3604       3604       3604       3604       3604       3604       3604      
   pop_diff.8 pop_diff.9 pop_diff.10
2  8572       8572       8572       
7  5211       5211       5211       
8  7343       7343       7343       
20  809        809        809       
25 8440       8440       8440       
27 7241       7241       7241       
30 1387       1387       1387       
35 9303       9303       9303       
36 6983       6983       6983       
42 7484       7484       7484       
46  522        522        522       
49  996        996        996       
51 3604       3604       3604

[1] 2

[1] 13